### PR TITLE
tests: some fixes and improvements for nested execution

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This change includes:
. Some tests could fail because logs dir not created when the vm is
started
. Creating environment for manual nested suite in spread.yaml
. Restoring environment for manual nested suite in spread.yaml
. Clean the extra-snaps directory as well

### CODE REVIEW

File: spread.yaml [hunk 1]
```diff
@@ -848,14 +848,15 @@ suites:
             distro_update_package_db
             distro_install_package snapd qemu qemu-utils genisoimage sshpass qemu-kvm cloud-image-utils ovmf kpartx xz-utils
             snap install ubuntu-image --classic
-
-            # TODO: replace with distro package once 2.45 is in stable
-            dpkg -i "$SPREAD_PATH"/../snapd_*.deb
```
Comment (anonymouse64) [inline spread.yaml:853]: hmm why is it desirable to just use snapd from the distro instead of the one we build for the spread run?

for example the distro package of snapd will not trust the testkeys which I am currently writing tests for, so if we do go back to using snapd from distro for the host VM, my test will need to install the snapd from the spread run and then uninstall it which is awkward and feels wrong

File: spread.yaml [hunk 2]
```diff
@@ -848,14 +848,15 @@ suites:
...
-
+        prepare-each: |
             #shellcheck source=tests/lib/nested.sh
             . "$TESTSLIB/nested.sh"
             nested_prepare_env
-
+        restore-each: |
```
Comment (anonymouse64) [inline spread.yaml:855]: hmm does restore-each get run _before_ the debug shell or after? I think it's after in which case this change is fine but I don't recall directly and the docs on spread don't clarify when exactly debug shells get started and I'm too lazy to go look at the code :smile:

**Review by anonymouse64 — COMMENTED** [review]
LGTM with the exception of one wondering about why we don't want to use snapd from the spread run anymore

File: spread.yaml (hunk 1 above)
Comment (sergiocazzolato) [reply spread.yaml:853]: Nice, so I reverted that and installed snapd in all the nested suites.
Thanks for the review

File: spread.yaml (hunk 2 above)
Comment (sergiocazzolato) [reply spread.yaml:858]: Correct, the restore happens after the debug.

File: spread.yaml [hunk 3]
```diff
@@ -1022,9 +1032,8 @@ suites:
             distro_install_package snapd qemu qemu-utils genisoimage sshpass qemu-kvm cloud-image-utils ovmf kpartx xz-utils
             snap install ubuntu-image --classic
 
-            # TODO: replace with distro package once 2.45 is in stable
+            # Install the snapd built
             dpkg -i "$SPREAD_PATH"/../snapd_*.deb
-            snap install core --beta
```
Comment (stolowski) [inline spread.yaml:1027]: Why this change?

File: spread.yaml [hunk 4]
```diff
@@ -849,13 +849,17 @@ suites:
...
+        prepare-each: |
             #shellcheck source=tests/lib/nested.sh
             . "$TESTSLIB/nested.sh"
             nested_prepare_env
-
+        restore-each: |
+            #shellcheck source=tests/lib/nested.sh
+            . "$TESTSLIB/nested.sh"
+            nested_destroy_vm
+            nested_cleanup_env
```
Comment (stolowski) [inline spread.yaml:862]: Nice, I like how it simplifies restore sections of the tests below 👍

File: spread.yaml [hunk 5]
```diff
@@ -962,6 +969,9 @@ suites:
             distro_install_package snapd qemu qemu-utils genisoimage sshpass qemu-kvm cloud-image-utils ovmf kpartx xz-utils
             snap install ubuntu-image --classic
 
+            # Install the snapd built
+            dpkg -i "$SPREAD_PATH"/../snapd_*.deb
```
Comment (stolowski) [inline spread.yaml:973]: Why do we need that on the host when testing core in nested?

**Review by stolowski — COMMENTED** [review]
Looks good overall, thanks for simplifying nested test cleanup, a few questions.

File: spread.yaml (hunk 4 above)
Comment (sergiocazzolato) [reply spread.yaml:862]: yes

File: spread.yaml (hunk 3 above)
Comment (sergiocazzolato) [reply spread.yaml:1027]: This is because at some point we added that because it was needed a change that was just in edge and beta, but it is not required now. I left comments in other suites about that.

File: spread.yaml (hunk 5 above)
Comment (anonymouse64) [reply spread.yaml:973]: Because for example we use the snapd on the host when we create the image, when we inject snap-bootstrap into the rebuilt kernel snap initramfs for uc20 and for any fakestore operations that I will be proposing, so it's more convenient and expected IMHO to get snapd from the spread run on the host when preparing the image etc.

Comment (stolowski) [reply spread.yaml:973]: Ah, I see, thanks for explanation!

**Comment by sergiocazzolato** [comment]
@anonymouse64 @stolowski thanks for the review
I added a new change to improve how nested helper wait until the reboot is completed.

File: tests/lib/nested.sh [hunk 6]
```diff
@@ -19,11 +19,42 @@ NESTED_SSH_PORT=8022
...
-    while ! nested_exec true; do
+    last_reboot=""
+    while [ $retry -ge 0 ]; do
+        retry=$(( retry - 1 ))
+        last_reboot="$(nested_get_last_reboot)"
+        if [[ "$last_reboot" =~ .*-.*-.*T.*:.*:.*Z ]] && [ "$last_reboot" != "$initial_reboot" ]; then
+            break
+        fi
+        sleep "$wait"
+    done
```
Comment (stolowski) [inline tests/lib/nested.sh]: Could this be simplified by using  /proc/sys/kernel/random/boot_id ?

Comment (sergiocazzolato) [reply tests/lib/nested.sh]: Nice, dind't know about this approach, I already updated that.
Thanks a lot

File: tests/lib/nested.sh [hunk 7]
```diff
@@ -26,29 +26,27 @@ nested_wait_for_no_ssh() {
...
     wait=5
 
-    last_reboot=""
+    last_boot_id=""
     while [ $retry -ge 0 ]; do
         retry=$(( retry - 1 ))
-        last_reboot="$(nested_get_last_reboot)"
-        if [[ "$last_reboot" =~ .*-.*-.*T.*:.*:.*Z ]] && [ "$last_reboot" != "$initial_reboot" ]; then
+        last_boot_id="$(nested_get_boot_id)"
+        if [[ "$last_boot_id" =~ .*-.*-.*-.*-.* ]] && [ "$last_boot_id" != "$initial_boot_id" ]; then
```
Comment (stolowski) [inline tests/lib/nested.sh:43]: The first part of the conditional should be removed now, it made sense only with uptime-based check, no?

Comment (stolowski) [reply tests/lib/nested.sh:43]: Ah, you updated it to match boot_id string, that's fine.

File: tests/lib/nested.sh [hunk 8]
```diff
@@ -19,11 +19,42 @@ NESTED_SSH_PORT=8022
...
 
-    while ! nested_exec true; do
+nested_wait_for_reboot() {
+    local initial_boot_id="$1"
+    local retry wait last_boot_id
+    retry=120
+    wait=5
+
+    last_boot_id=""
+    while [ $retry -ge 0 ]; do
```
Comment (anonymouse64) [inline tests/lib/nested.sh:40]: mmm so in the worst case that we can't ssh to the machine (which is all too common unfortunately :frowning: ), this code results in waiting for a total of `120*(200*1 + 5) = 24600` seconds or almost 7 hours, which I think is a tad too long for this kind of timeout, can we instead something fail quicker?

File: tests/lib/nested.sh [hunk 9]
```diff
@@ -33,15 +64,15 @@ nested_wait_for_ssh() {
...
-    local retry wait
-    retry=200
-    wait=1
+nested_retry_until_success() {
+    local retry="$1"
+    local wait="$2"
+    shift 2
 
-    while nested_exec true; do
+    while ! nested_exec "$@"; do
```
Comment (anonymouse64) [inline tests/lib/nested.sh:72]: why not just `until` directly? 

```suggestion
    until nested_exec "$@"; do
```

**Review by anonymouse64 — COMMENTED** [review]
Thanks for the fixes to use the boot id, I think that's a definite improvement over the previous implementation, the only concern I have now about this PR is about the excessive timeout in the worst case where we can't SSH to the machine anymore, as I think almost an entire workday is too long to wait for a test to fail :smile:

File: tests/lib/nested.sh (hunk 8 above)
Comment (sergiocazzolato) [reply tests/lib/nested.sh:40]: yes, I'll try.

Comment (sergiocazzolato) [reply tests/lib/nested.sh:40]: @anonymouse64 I just pushed a fix for that. Thanks for the review

**Comment by sergiocazzolato** [comment]
@anonymouse64 @stolowski commets addressed. 
Thanks

**Review by mvo5 — APPROVED** [review]
I merged master now and removed one redundant change that was already on master. If the nested execution is green I'm strongly in favor of merging, we can always cleanup further but right now we have failures in nested on master and this PR appears to fix them). This is also why I added the criticial tag.

**Review by stolowski — APPROVED** [review]
Thanks for the improvements and changes, detecting reboots looks more robust now, and also cleanup is nicer. It's a bit hard to assess every detail of nested-related changes and I think a spread run is an ultimate confirmation, so if it is green, I'm happy to land it!
